Dokka Gradle plugin doc update (#4354)
orangy/dokka: docs/topics/dokka-gradle-configuration-options.md

See [Configuration examples](dokka-gradle.md#configuration-examples) for more details on where to apply configuration blocks and how.

### General configuration

Here is an example of the general Dokka Gradle plugin configuration. Use the top-level `dokka {}` DSL configuration.

In DGP, `dokkaPublications` is the central way to declare Dokka publication configurations.
The default publications
are [`html`](dokka-html.md) and [`javadoc`](dokka-javadoc.md).

The syntax of `build.gradle.kts` files differs from regular `.kt`
files (such as those used for custom Gradle plugins) because Gradle's Kotlin DSL uses type-safe accessors:

<tabs group="build-script">
<tab title="Gradle Kotlin DSL" group-key="kotlin">

```kotlin
plugins {
    id("org.jetbrains.dokka") version "%dokkaVersion%"
}

dokka {
    dokkaPublications.html {
        moduleName.set(project.name)
        moduleVersion.set(project.version.toString())
        outputDirectory.set(layout.buildDirectory.dir("dokka/html"))
        failOnWarning.set(false)
        suppressInheritedMembers.set(false)
        offlineMode.set(false)
        suppressObviousFunctions.set(true)

>         includes.from(project.files(), "packages.md", "extra.md")

Reviewer: ```suggestion
        includes.from("packages.md", "extra.md")
```

`project.files()` returns an empty files collection, so it can be removed.

        outputDirectory.set(rootDir.resolve("docs/api/0.x"))
        includes.from(project.layout.projectDirectory.file("README.md"))
    }
}
```

</tab>
<tab title="Kotlin custom plugin" group-key="kotlin custom">

```kotlin
// CustomPlugin.kt

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.jetbrains.dokka.gradle.DokkaExtension

abstract class CustomPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        project.plugins.apply("org.jetbrains.dokka")

        project.extensions.configure(DokkaExtension::class.java) { dokka ->
            
            dokka.moduleName.set(project.name)
            dokka.moduleVersion.set(project.version.toString())

            dokka.dokkaPublications.named("html") { publication ->
                // Standard output directory for HTML documentation
                publication.outputDirectory.set(project.layout.buildDirectory.dir("dokka/html"))
                publication.failOnWarning.set(true)
                publication.suppressInheritedMembers.set(true)
                publication.offlineMode.set(false)
                publication.suppressObviousFunctions.set(true)

>                 publication.includes.from(project.files(), "packages.md", "extra.md")

Reviewer: ```suggestion
                publication.includes.from("packages.md", "extra.md")
```

            }
            externalDocumentationLinks {
                // External documentation links section
            }
        }
    }
}
```

</tab>
<tab title="Gradle Groovy DSL" group-key="groovy">

```groovy
import org.jetbrains.dokka.gradle.engine.parameters.VisibilityModifier

dokka {
    // ..
    // General configuration section
    // ..

    dokkaSourceSets {
        // Example: Configuration exclusive to the 'linux' source set
        named("linux") {
            dependentSourceSets { named("native") }
            sourceRoots.from(file("linux/src"))
        }

        configureEach {
            suppress.set(false)
            displayName.set(name)
            documentedVisibilities.set([VisibilityModifier.Public] as Set) // OR documentedVisibilities(VisibilityModifier.Public)
            reportUndocumented.set(false)
            skipEmptyPackages.set(true)
            skipDeprecated.set(false)
            suppressGeneratedFiles.set(true)
            jdkVersion.set(8)
            languageVersion.set("1.7")
            apiVersion.set("1.7")
            sourceRoots.from(file("src"))
            classpath.from(files(), file("libs/dependency.jar"))

>             samples.from(files(), "samples/Basic.kt", "samples/Advanced.kt")

Reviewer: ```suggestion
            classpath.from(file("libs/dependency.jar"))
            samples.from("samples/Basic.kt", "samples/Advanced.kt")
```

        moduleVersion.set(project.version.toString())
        outputDirectory.set(layout.buildDirectory.dir("dokka/html"))
        failOnWarning.set(false)
        suppressInheritedMembers.set(false)
        offlineMode.set(false)
        suppressObviousFunctions.set(true)
        includes.from(project.files(), "packages.md", "extra.md")
   }

    dokkaSourceSets {
        // Example: Configuration exclusive to the 'linux' source set
        named("linux") {
            dependentSourceSets{named("native")}
            sourceRoots.from(file("linux/src"))
        }

        configureEach {
            suppress.set(false)
            displayName.set(name)
            documentedVisibilities.set(setOf(VisibilityModifier.Public)) // OR documentedVisibilities(VisibilityModifier.Public)
            reportUndocumented.set(false)
            skipEmptyPackages.set(true)
            skipDeprecated.set(false)
            suppressGeneratedFiles.set(true)
            jdkVersion.set(8)
            languageVersion.set("1.7")
            apiVersion.set("1.7")
            sourceRoots.from(file("src"))
            classpath.from(project.files(), file("libs/dependency.jar"))

>             samples.from(project.files(), "samples/Basic.kt", "samples/Advanced.kt")

Reviewer: ```suggestion
            classpath.from(file("libs/dependency.jar"))
            samples.from("samples/Basic.kt", "samples/Advanced.kt")
```

                matchingRegex.set(".*api.*")
                suppress.set(false)
                skipDeprecated.set(false)
                reportUndocumented.set(false)
                documentedVisibilities.set(
                    setOf(
                        VisibilityModifier.Public,
                        VisibilityModifier.Private,
                        VisibilityModifier.Protected,
                        VisibilityModifier.Internal,
                        VisibilityModifier.Package
                    )
                )
            }
        }
    }
}
```

</tab>
<tab title="Gradle Groovy DSL" group-key="groovy">

```groovy
import org.jetbrains.dokka.gradle.engine.parameters.VisibilityModifier

plugins {
    id 'org.jetbrains.dokka' version '%dokkaVersion%'
}

dokka {
    dokkaPublications {
        html {
            moduleName.set(project.name)
            moduleVersion.set(project.version.toString())
            outputDirectory.set(layout.buildDirectory.dir("dokka/html"))
            failOnWarning.set(false)
            suppressInheritedMembers.set(false)
            offlineMode.set(false)
            suppressObviousFunctions.set(true)

>             includes.from(files(), "packages.md", "extra.md")

Reviewer: ```suggestion
            includes.from("packages.md", "extra.md")
```

[//]: # (title: Dokka Gradle configuration options)

> Dokka has many configuration options to tailor your and your reader's experience.

Reviewer 2: ```suggestion
Dokka has many configuration options to customize your and your reader's experience.
```

This meaning of "tailor" usually is used with something we're tailoring stuff _to_.

Below are some examples and detailed descriptions for each configuration section. You can also find an example
with [all configuration options](#complete-configuration) applied.

> See [Configuration examples](dokka-gradle.md#configuration-examples) for more details on where to apply configuration blocks and how.

Reviewer 2: I'm a bit confused as to how the link's contents are different from the current page. Isn't this supposed to be the place with details and examples? Could we clarify this so that users can judge better whether to follow the link?

> See [Configuration examples](dokka-gradle.md#configuration-examples) for more details on where to apply configuration blocks and how.

Reviewer 2: ```suggestion
See [Configuration examples](dokka-gradle.md#configuration-examples) for more details on where and how to apply configuration blocks.
```

### General configuration

Here is an example of the general Dokka Gradle plugin configuration. Use the top-level `dokka {}` DSL configuration.

In DGP, `dokkaPublications` is the central way to declare Dokka publication configurations.
The default publications
are [`html`](dokka-html.md) and [`javadoc`](dokka-javadoc.md).

The syntax of `build.gradle.kts` files differs from regular `.kt`
files (such as those used for custom Gradle plugins) because Gradle's Kotlin DSL uses type-safe accessors:

<tabs group="build-script">
<tab title="Gradle Kotlin DSL" group-key="kotlin">

```kotlin
plugins {
    id("org.jetbrains.dokka") version "%dokkaVersion%"
}

dokka {
    dokkaPublications.html {
        moduleName.set(project.name)
        moduleVersion.set(project.version.toString())
        outputDirectory.set(layout.buildDirectory.dir("dokka/html"))
        failOnWarning.set(false)
        suppressInheritedMembers.set(false)
        offlineMode.set(false)
        suppressObviousFunctions.set(true)
        includes.from("packages.md", "extra.md")
        
        // Output directory for additional files

>         // Use this block when you want to change the output directory and include extra files

Reviewer 2: This line is cut off in the code block on a published page.

                // Output directory for additional files
                // Use this block when you want to change the output directory and include extra files
                html.outputDirectory.set(project.rootDir.resolve("docs/api/0.x"))
                html.includes.from(project.layout.projectDirectory.file("README.md"))
            }
        }
    }
}
```

</tab>
<tab title="Gradle Groovy DSL" group-key="groovy">

```groovy
plugins {
    id 'org.jetbrains.dokka' version '%dokkaVersion%'
}

dokka {
    dokkaPublications {
        html {
            // Sets general module information
            moduleName.set(project.name)
            moduleVersion.set(project.version.toString())

            // Standard output directory for HTML documentation
            outputDirectory.set(layout.buildDirectory.dir("dokka/html"))

            // Core Dokka options
            failOnWarning.set(false)
            suppressInheritedMembers.set(false)
            offlineMode.set(false)
            suppressObviousFunctions.set(true)
            includes.from(files("packages.md", "extra.md"))

            // Output directory for additional files

>             // Use this block when you want to change the output directory and include extra files

Reviewer 2: Also cut off on the page. (For "Kotlin custom plugin" doesn't matter as much because lines are long in general.)

            outputDirectory.set(file("$rootDir/docs/api/0.x"))
            includes.from(layout.projectDirectory.file("README.md"))
        }
    }
}
```

</tab>
</tabs>

<deflist collapsible="true">
    <def title="moduleName">
        <p>
           The display name to refer to the project’s documentation. It appears in the table of contents, navigation, 

>            headers, and log messages. In multi-project builds, each subproject <code>moduleName</code> is 

Reviewer 2: ```suggestion
           headers, and log messages. In multi-project builds, each subproject's <code>moduleName</code> is 
```

        <p>This setting applies to all documentation formats (HTML, Javadoc, etc.) generated by the <code>dokkaGenerate</code> task.</p>
        <p>Default: <code>build/dokka/html</code></p>
        <p><b>Output directory for additional files</b></p>
        <p>You can specify the output directory and include additional files for both single and multi-project builds.
           For multi-project builds,
           set the output directory and include additional files (such as <code>README.md</code>) 
           in the configuration of the root project.
        </p>
    </def>
    <def title="failOnWarning">
        <p>
            Determines whether Dokka should fail the build when a warning occurs during documentation generation.
            The process waits until all errors and warnings have been emitted first.
        </p>
        <p>This setting works well with <code>reportUndocumented</code>.</p>
        <p>Default: <code>false</code></p>
    </def>
    <def title="suppressInheritedMembers">
        <p>Whether to suppress inherited members that aren't explicitly overridden in a given class.</p>
        <p>

>             Note: This can suppress functions such as <code>equals</code> / <code>hashCode</code> / <code>toString</code>, 

Reviewer 2: ```suggestion
            Note: This suppresses functions such as <code>equals</code> / <code>hashCode</code> / <code>toString</code>, 
```

            but cannot suppress synthetic functions such as <code>dataClass.componentN</code> and 
            <code>dataClass.copy</code>. Use <code>suppressObviousFunctions</code>
            for that.
        </p>
        <p>Default: <code>false</code></p>
    </def>
    <def title="offlineMode">

>         <p>Whether to resolve remote files/links over your network.</p>

Reviewer 2: ```suggestion
        <p>Whether to resolve remote files and links over your network.</p>
```

        <p>Default: <code>false</code></p>
    </def>
    <def title="suppressObviousFunctions">
        <p>Whether to suppress obvious functions.</p>
        <p>
            A function is considered to be obvious if it is:</p>
            <list>
                <li>
                    Inherited from <code>kotlin.Any</code>, <code>Kotlin.Enum</code>, <code>java.lang.Object</code> or
                    <code>java.lang.Enum</code>, such as <code>equals</code>, <code>hashCode</code>, <code>toString</code>.
                </li>
                <li>
                    Synthetic (generated by the compiler) and does not have any documentation, such as
                    <code>dataClass.componentN</code> or <code>dataClass.copy</code>.
                </li>
            </list>
        <p>Default: <code>true</code></p>
    </def>
     <def title="includes">
        <p>
            A list of Markdown files that contain
            <a href="dokka-module-and-package-docs.md">subproject and package documentation</a>.
        </p>
        <p>The contents of the specified files are parsed and embedded into documentation as subproject and package descriptions.</p>
        <p>

>             See <a href="https://github.com/Kotlin/dokka/blob/master/examples/gradle-v2/basic-gradle-example/build.gradle.kts">Dokka gradle example</a>

Reviewer 2: ```suggestion
            See <a href="https://github.com/Kotlin/dokka/blob/master/examples/gradle-v2/basic-gradle-example/build.gradle.kts">Dokka Gradle example</a>
```

            samples.from("samples/Basic.kt", "samples/Advanced.kt")

            sourceLink {
                // Source link section
            }
            perPackageOption {
                // Package options section
            }
            externalDocumentationLinks {
                // External documentation links section
            }
        }
    }
}
```

</tab>
</tabs>

<deflist collapsible="true">
    <def title="suppress">
        <p>Whether this source set should be skipped when generating documentation.</p>
        <p>Default: <code>false</code></p>
    </def>
    <def title="displayName">
        <p>The display name used to refer to this source set.</p>
        <p>
            The name is used both externally (for example, as source set name visible to documentation readers) and 
            internally (for example, for logging messages of <code>reportUndocumented</code>).
        </p>
        <p>By default, the value is deduced from information provided by the Kotlin Gradle plugin.</p>
    </def>
    <def title="documentedVisibilities">
        <p>Defines which visibility modifiers Dokka should include in the generated documentation.</p>
        <p>

>             Use them if you want to document <code>Protected</code>/<code>Internal</code>/<code>Private</code> declarations,

Author: I think they are in uppercase, for example: `documentedVisibilities.set(setOf(VisibilityModifier.Public))`

        <p>Whether to document declarations annotated with <code>@Deprecated</code>.</p>
        <p>This can be configured on a per-package basis.</p>
        <p>Default: <code>false</code></p>
    </def>
    <def title="suppressGeneratedFiles">
        <p>Whether to document/analyze generated files.</p>
        <p>
            Generated files are expected to be present under the <code>{project}/{buildDir}/generated</code> directory.
        </p>
        <p>
            If set to <code>true</code>, it effectively adds all files from that directory to the
            <code>suppressedFiles</code> option, so you can configure it manually.
        </p>
        <p>Default: <code>true</code></p>
    </def>
    <def title="jdkVersion">
        <p>The JDK version to use when generating external documentation links for Java types.</p>
        <p>
            For example, if you use <code>java.util.UUID</code> in some public declaration signature,
            and this option is set to <code>8</code>, Dokka generates an external documentation link
            to <a href="https://docs.oracle.com/javase/8/docs/api/java/util/UUID.html">JDK 8 Javadocs</a> for it.
        </p>

>         <p>Default: JDK 8</p>

Reviewer 2: ```suggestion
        <p>Default: 8</p>
```

According to the example above.

            environment.
        </p>
        <p>By default, the latest language version available to Dokka's embedded compiler is used.</p>
    </def>
    <def title="apiVersion">
        <p>
            <a href="https://kotlinlang.org/docs/compatibility-modes.html">The Kotlin API version</a>
            used for setting up analysis and <a href="https://kotlinlang.org/docs/kotlin-doc.html#sample-identifier">@sample</a>
            environment.
        </p>
        <p>By default, it is deduced from <code>languageVersion</code>.</p>
    </def>
    <def title="sourceRoots">
        <p>
            The source code roots to be analyzed and documented.
            Acceptable inputs are directories and individual <code>.kt</code> / <code>.java</code> files.
        </p>
        <p>By default, source roots are deduced from information provided by the Kotlin Gradle plugin.</p>
    </def>
    <def title="classpath">
        <p>The classpath for analysis and interactive samples.</p>
        <p>This is useful if some types that come from dependencies are not resolved/picked up automatically.</p>
        <p>This option accepts both <code>.jar</code> and <code>.klib</code> files.</p>

>         <p>By default, the classpath is deduced from information provided by the Kotlin Gradle plugin.</p>

Reviewer 2: This is not critical, but sometimes the default is introduced with "Default:" and sometimes with "By default", which could be more uniform.

    </def>
    <def title="samples">
        <p>
            A list of directories or files that contain sample functions which are referenced via the
            <a href="https://kotlinlang.org/docs/kotlin-doc.html#sample-identifier">@sample</a> KDoc tag.
        </p>
    </def>
</deflist>

### Source link configuration

Configure source links
to allow navigation from the generated documentation to the corresponding source code in a remote repository.
Use the `dokkaSourceSets.main{}` block for this configuration.

The `sourceLinks` configuration block allows you to add a `source` link to each signature
that leads to the `remoteUrl` with a specific line number.
The line number is configurable by setting `remoteLineSuffix`.

> This helps readers to find the source code for each declaration.

Reviewer 2: This sentence can be embedded in the first paragraph: "Configure source links to help readers find the source for each declaration in a remote repository." which is more to the point IMO.

            perPackageOption {
                matchingRegex.set(".*api.*")
                suppress.set(false)
                skipDeprecated.set(false)
                reportUndocumented.set(false)
                documentedVisibilities.set(setOf(VisibilityModifier.Public)) // OR documentedVisibilities(VisibilityModifier.Public)
            }
        }
    }
}
```

</tab>
<tab title="Gradle Groovy DSL" group-key="groovy">

```groovy
import org.jetbrains.dokka.gradle.engine.parameters.VisibilityModifier

dokka {
    dokkaPublications {
        html {
            dokkaSourceSets.configureEach {
                perPackageOption {
                    matchingRegex.set(".*api.*")
                    suppress.set(false)
                    skipDeprecated.set(false)
                    reportUndocumented.set(false)
                    documentedVisibilities.set([VisibilityModifier.Public] as Set)
                }
            }
        }
    }
}
```

</tab>
</tabs>

<deflist collapsible="true">
    <def title="matchingRegex">
        <p>The regular expression that is used to match the package.</p>
        <p>Default: <code>.*</code></p>
    </def>
    <def title="suppress">

>         <p>Whether this package should be skipped when generating documentation.</p>

Reviewer 2: ```suggestion
        <p>Whether the package matching the `matchingRegex` should be skipped when generating documentation.</p>
```

Is this correct? "This" seems vague.

        <p>Default: <code>false</code></p>
    </def>
    <def title="skipDeprecated">
        <p>Whether to document declarations annotated with <code>@Deprecated</code>.</p>
        <p>This can be configured on the source set level.</p>
        <p>Default: <code>false</code></p>
    </def>
    <def title="reportUndocumented">
        <p>
            Whether to emit warnings about visible undocumented declarations, that is declarations without KDocs
            after they have been filtered by <code>documentedVisibilities</code> and other filters.
        </p>
        <p>This setting works well with <code>failOnWarning</code>.</p>
        <p>This can be configured on the source set level.</p>
        <p>Default: <code>false</code></p>
    </def>
    <def title="documentedVisibilities">
        <p>Defines which visibility modifiers Dokka should include in the generated documentation.</p>
        <p>
            Use them if you want to document <code>Protected</code>/<code>Internal</code>/<code>Private</code> 
            declarations within this package,

>             as well as if you want to exclude <code>Public</code> declarations and only document internal API.

Reviewer 2: ```suggestion
            as well as if you want to exclude <code>public</code> declarations and only document internal API.
```
